### PR TITLE
Liquibase changes

### PR DESCRIPTION
**Pull request by patryllus**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/muzima/handler/JsonRegistrationQueueDataHandler.java [hunk 1]
```diff
@@ -65,6 +66,18 @@
     Set<PersonAttribute> personAttributes;
     private QueueProcessorException queueProcessorException;
 
+    public static final String NEXT_OF_KIN_ADDRESS = "7cf22bec-d90a-46ad-9f48-035952261294";
```
Comment (benamonya) [inline api/src/main/java/org/openmrs/module/muzima/handler/JsonRegistrationQueueDataHandler.java:67]: Hardcoded concept UUID not acceptable since not all implementations will have them.

File: api/src/main/java/org/openmrs/module/muzima/handler/JsonRegistrationQueueDataHandler.java [hunk 2]
```diff
@@ -151,10 +164,11 @@ private void setPatientIdentifiersFromPayload() {
     }
 
     private PatientIdentifier getPreferredPatientIdentifierFromPayload(){
-        String identifierValue = JsonUtils.readAsString(payload, "$['patient']['patient.medical_record_number']");
-        String identifierTypeName = "AMRS Universal ID";
+//        String identifierValue = JsonUtils.readAsString(payload, "$['patient']['patient.medical_record_number']");
```
Comment (benamonya) [inline api/src/main/java/org/openmrs/module/muzima/handler/JsonRegistrationQueueDataHandler.java:165]: Removing AMRS Universal ID will break ampath implementations. This handler was made specifically for ampath, instead you should use https://github.com/muzima/openmrs-module-muzimacore/blob/master/api/src/main/java/org/openmrs/module/muzima/handler/JsonGenericRegistrationQueueDataHandler.java

File: api/src/main/java/org/openmrs/module/muzima/handler/JsonRegistrationQueueDataHandler.java [hunk 3]
```diff
@@ -362,4 +420,23 @@ private Patient findSimilarSavedPatient() {
     public boolean accept(final QueueData queueData) {
         return StringUtils.equals(DISCRIMINATOR_VALUE, queueData.getDiscriminator());
     }
+    /**
+     * Can't save patients unless they have required OpenMRS IDs
```
Comment (benamonya) [inline api/src/main/java/org/openmrs/module/muzima/handler/JsonRegistrationQueueDataHandler.java:421]: Making this mandatory will be problematic to most implementation since they may not be having this identifier. The UUID is also hardcoded thus not generic.

File: api/src/main/resources/liquibase.xml [hunk 4]
```diff
@@ -25,6 +25,17 @@
             <column name="name" type="varchar(255)"/>
         </createTable>
     </changeSet>
+    <changeSet id="muzima-20170828-1811" author="aojwang">
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:33]: This changeset needs discussion since not all implementations will be using UTF-8 as default database charset.

File: api/src/main/resources/liquibase.xml [hunk 5]
```diff
@@ -525,6 +619,9 @@
             <column name="sender" type="int">
                 <constraints nullable="false"/>
             </column>
+            <column name="patient" type="int">
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:759]: Why is the patient required since notifications can be client-specific or general

File: api/src/main/resources/liquibase.xml [hunk 6]
```diff
@@ -1184,6 +1335,31 @@
         </insert>
     </changeSet>
 
+    <!--Adding scheduled task to periodically refresh ETL tables-->
+    <changeSet id="${project.parent.artifactId}-20180802-1220" author="aojwang">
+        <preConditions onFail="MARK_RAN">
+            <sqlCheck expectedResult="0">
+                SELECT COUNT(*) FROM scheduler_task_config
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:1513]: Is the precondition supposed to check for RefreshETLTablesTask or for ProcessQueueDataTask. Also, consider changing the changeset id.

File: api/src/main/resources/liquibase.xml [hunk 7]
```diff
@@ -1184,6 +1335,31 @@
...
+                And name = 'Refresh ETL Tables'
+            </sqlCheck>
+        </preConditions>
+        <comment>Inserting Muzima Queue Data Processor into 'schedule_task_config' table</comment>
+        <insert tableName="scheduler_task_config">
+            <column name="name" value="Muzima Queue Data Processor" />
+            <column name="description" value="Processes Muzima Queue Data Processor" />
+            <column name="schedulable_class" value="org.openmrs.module.muzima.task.ProcessQueueDataTask" />
+            <column name="start_time_pattern" value="MM/dd/yyyy HH:mm:ss" />
+            <column name="start_time" valueDate="2017-11-28T23:59:59" />
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:1524]: Change start time to the current timestamp.

File: api/src/main/java/org/openmrs/module/muzima/handler/JsonEncounterQueueDataHandler.java [hunk 8]
```diff
@@ -121,19 +125,8 @@ public void process(final QueueData queueData) throws QueueProcessorException {
 
         try {
             if (validate(queueData)) {
-                Set<Obs> obss =  encounter.getAllObs();
```
Comment (benamonya) [inline api/src/main/java/org/openmrs/module/muzima/handler/JsonEncounterQueueDataHandler.java]: Why was this code block removed? This was meant to check to ensure all obs are valid before passing them to openmrs.

File: api/src/main/java/org/openmrs/module/muzima/handler/JsonEncounterQueueDataHandler.java [hunk 9]
```diff
@@ -290,18 +283,7 @@ private void createObs(final Encounter encounter, final Obs parentObs, final Con
                 Date obsDateTime = parseDate(dateString);
                 obs.setObsDatetime(obsDateTime);
             }
-        }else if(o instanceof JSONObject){
```
Comment (benamonya) [inline api/src/main/java/org/openmrs/module/muzima/handler/JsonEncounterQueueDataHandler.java]: Why is this code block deleted? This is supposed to cater for obs that come with obs datetime eg lab tests

File: api/src/main/java/org/openmrs/module/muzima/handler/JsonEncounterQueueDataHandler.java [hunk 10]
```diff
@@ -440,22 +421,16 @@ private void processEncounter(final Encounter encounter, final Object encounterO
...
 
     /**
-     * 
+     *
      * @param dateValue - String representation of the date value.
      * @return java.util.Date Object
      */
     private Date parseDate(final String dateValue) {
         Date date = null;
-        if(StringUtils.isNumeric(dateValue))
```
Comment (benamonya) [inline api/src/main/java/org/openmrs/module/muzima/handler/JsonEncounterQueueDataHandler.java]: What necessitated this deletion. We expect either date or numeric(timestamp) values for this method. If it is a numeric value, it cant be processed by dateFormat.parse().

File: api/src/main/java/org/openmrs/module/muzima/handler/JsonEncounterQueueDataHandler.java [hunk 11]
```diff
@@ -465,12 +440,100 @@ public boolean accept(final QueueData queueData) {
         return StringUtils.equals(DISCRIMINATOR_VALUE, queueData.getDiscriminator());
     }
 
+    protected static void useNewVisit(Encounter encounter) {
+        String VISIT_SOURCE_FORM = "8bfab185-6947-4958-b7ab-dfafae1a3e3d";
```
Comment (benamonya) [inline api/src/main/java/org/openmrs/module/muzima/handler/JsonEncounterQueueDataHandler.java:444]: Hardcoded UUID not acceptable

File: api/src/main/java/org/openmrs/module/muzima/handler/JsonEncounterQueueDataHandler.java [hunk 12]
```diff
@@ -465,12 +440,100 @@ public boolean accept(final QueueData queueData) {
...
+        Visit visit = new Visit();
+        visit.setStartDatetime(OpenmrsUtil.firstSecondOfDay(encounter.getEncounterDatetime()));
+        visit.setStopDatetime(OpenmrsUtil.getLastMomentOfDay(encounter.getEncounterDatetime()));
+        visit.setLocation(encounter.getLocation());
+        visit.setPatient(encounter.getPatient());
+        visit.setVisitType(Context.getVisitService().getVisitTypeByUuid("3371a4d4-f66f-4454-a86d-92c7b3da990c"));
```
Comment (benamonya) [inline api/src/main/java/org/openmrs/module/muzima/handler/JsonEncounterQueueDataHandler.java:450]: Hardcoded UUID not acceptable

File: api/src/main/java/org/openmrs/module/muzima/handler/JsonEncounterQueueDataHandler.java [hunk 13]
```diff
@@ -465,12 +440,100 @@ public boolean accept(final QueueData queueData) {
...
+        // Set to new visit
+        encounter.setVisit(visit);
+
+        if (visit != null) {
+            visit.addEncounter(encounter);
+        }
+    }
+
     /**
-     * Checks if all obs have concepts with valid datatype.
```
Comment (benamonya) [inline api/src/main/java/org/openmrs/module/muzima/handler/JsonEncounterQueueDataHandler.java]: What was the reason of removing this code block?

File: api/src/main/resources/liquibase.xml [hunk 14]
```diff
@@ -490,17 +473,19 @@
         </sql>
     </changeSet>
 
-    <changeSet id="muzimaform-2014-10-31" author="SahajSoft">
+    <changeSet id="muzimaform-2019-08-07" author="pwangoo">
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:476]: Why are the changeset id and author changed here?

File: api/src/main/resources/liquibase.xml [hunk 15]
```diff
@@ -1268,7 +1249,7 @@
         </sql>
     </changeSet>
 
-    <changeSet id="muzima-20160901-093538" author="smbugua">
+    <changeSet id="muzima-20190808-095338" author="pwangoo">
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:1252]: Why the change of changeset id and author.

File: api/src/main/resources/liquibase.xml [hunk 16]
```diff
@@ -1335,10 +1313,10 @@
                                  referencedTableName="users" referencedColumnNames="user_id"/>
     </changeSet>
 
-    <changeSet id="muzima-20160901-115152" author="smbugua">
+    <changeSet id="muzima-20190806-150052" author="pwangoo">
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:1316]: Why the change of changeset id and author.

File: api/src/main/resources/liquibase.xml [hunk 17]
```diff
@@ -1406,20 +1384,21 @@
         </sql>
     </changeSet>
 
-    <changeSet id="muzima-20160922-145252" author="smbugua">
+    <changeSet id="muzima-20190808-101052" author="pwangoo">
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:1387]: Why the change of changeset id and author.

File: api/src/main/resources/liquibase.xml [hunk 18]
```diff
@@ -1335,10 +1313,10 @@
                                  referencedTableName="users" referencedColumnNames="user_id"/>
     </changeSet>
 
-    <changeSet id="muzima-20160901-115152" author="smbugua">
+    <changeSet id="muzima-20190806-150052" author="pwangoo">
         <preConditions onFail="MARK_RAN">
             <and>
-                <tableExists tableName="muzima_form_tag"/>
+                <columnExists tableName="muzima_form_tag" columnName="muzimaforms_tag_id"/>
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:1319]: Shouldn't we have a check if table exists then check for the column

File: api/src/main/resources/liquibase.xml [hunk 19]
```diff
@@ -1481,7 +1460,10 @@
         </sql>
     </changeSet>
 
-    <changeSet id="muzima-20170922145600" author="mssavai">
+    <changeSet id="muzima-20190813185600" author="pwangoo">
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:1463]: Why the change of changeset id and author.

File: api/src/main/resources/liquibase.xml [hunk 20]
```diff
@@ -1494,7 +1476,10 @@
         </insert>
     </changeSet>
 
-    <changeSet id="muzima-20171017151600" author="mssavai">
+    <changeSet id="muzima-20190813185700" author="pwangoo">
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:1479]: Why the change of changeset id and author

File: api/src/main/resources/liquibase.xml [hunk 21]
```diff
@@ -1507,7 +1492,10 @@
         </insert>
     </changeSet>
 
-    <changeSet id="muzima-20180214100000" author="bmokaya">
+    <changeSet id="muzima-20190813185800" author="pwangoo">
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:1495]: Why the change of changeset id and author.

File: api/src/main/resources/liquibase.xml [hunk 22]
```diff
@@ -1543,11 +1531,26 @@
         </insert>
     </changeSet>
 
-    <changeSet id="muzima-20180720102200" author="bmokaya">
+    <changeSet id="muzima-20190807172200" author="pwangoo">
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:1534]: Why the change of changeset id and author.

File: api/src/main/resources/liquibase.xml [hunk 23]
```diff
@@ -1543,11 +1531,26 @@
...
+                FROM  INFORMATION_SCHEMA.COLUMNS
+                WHERE table_name='muzima_notification_data'
+                AND column_name='patient'
+            </sqlCheck>
+        </preConditions>
         <dropNotNullConstraint columnDataType="int" columnName="patient" tableName="muzima_notification_data"/>
     </changeSet>
 
-    <changeSet id="muzima-20180717115600" author="mssavai">
+    <changeSet id="muzima-20190813265900" author="pwangoo">
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:1557]: Why the change of changeset id and author. Please also consider all the changes below this comment.

File: api/src/main/resources/liquibase.xml [hunk 24]
```diff
@@ -1789,4 +1832,48 @@
         </insert>
     </changeSet>
 
+    <changeSet id="muzima-20190808172200" author="pwangoo">
```
Comment (benamonya) [inline api/src/main/resources/liquibase.xml:1843]: Why the drop of the null constraint for patientUuid for the 3 tables.

**Comment by benamonya** [comment]
Hi @patryllus, Please resolve the conflicts and address the code review comments raised.

**Comment by patryllus** [comment]
Hi
 Hey Benard, I will be reviewing these and providing feedback with regard
with the comments provided.
